### PR TITLE
Surface asset check run failures

### PR DESCRIPTION
**Pull request by johannkm**
Per the write path described in https://github.com/dagster-io/dagster/pull/16006, the status goes

- planned event
- row created with status PLANNED
- evaluation event
- row updated with status SUCCESS/FAILURE

but there's also a case where

- planned event
- row created with status PLANNED
- run failure

and the check stays in PLANNED status. Similarly to how we reconcile asset status at read time, use the status of the associated run to adjust the status when we surface it over gql.

---

In the future with both assets and asset checks, I'd like to move to emitting explicit MATERIALIZATION_FAILURE/ASSET_CHECK_RUN_FAILURE events when the step fails (or when the run monitoring daemon fails the run). Plus analogous skip or cancel events. This would mean
- we can keep actual status in the db. helpful for filtering on status etc.
- we can use the event log consumer to alert on materialization and check run failures

but that's a bit larger of a project, and I think it's safe to mimic asset behavior in the meantime

### CODE REVIEW

**Comment by johannkm** [comment]
Current dependencies on/for this PR:
* master
  * **PR #15963** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15963" target="_blank">Graphite</a> 
    * **PR #16006** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16006" target="_blank">Graphite</a> 
      * **PR #16063** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16063" target="_blank">Graphite</a> 
        * **PR #16091** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16091" target="_blank">Graphite</a> 
          * **PR #16079** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16079" target="_blank">Graphite</a> 
            * **PR #16102** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16102" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16102?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/asset_layer.py [hunk 1]
```diff
@@ -687,6 +687,23 @@ def asset_info_for_output(
     ) -> Optional[AssetOutputInfo]:
         return self.asset_info_by_node_output_handle.get(NodeOutputHandle(node_handle, output_name))
 
+    def asset_check_handle_for_output(
```
[the remark this answers is not in the document]
Comment (johannkm) [reply python_modules/dagster/dagster/_core/definitions/asset_layer.py]: Ah this diff ended up with a commit from lower in the stack that already landed: https://github.com/dagster-io/dagster/pull/16091

I can improve this in a separate diff though

**Comment by johannkm** [comment]
I updated to an approach I think is more clear:

Separate enums, AssetCheckExecutionStoredStatus and AssetCheckExecutionResolvedStatus. Stored status is what we hold in the DB, and resolved status is what we get at read time when we check if a row with PLANNED status is actually still executing.

File: python_modules/dagster/dagster/_core/storage/asset_check_execution_record.py [hunk 2]
```diff
@@ -7,7 +7,9 @@
 class AssetCheckExecutionStatus(enum.Enum):
     PLANNED = "PLANNED"
     SUCCESS = "SUCCESS"
-    FAILURE = "FAILURE"
+    FAILURE = "FAILURE"  # explicit fail result
```
[the remark this answers is not in the document]
Comment (johannkm) [reply python_modules/dagster/dagster/_core/storage/asset_check_execution_record.py:14]: Updated by splitting the enums

File: python_modules/dagster/dagster/_core/storage/asset_check_execution_record.py [hunk 3]
```diff
@@ -4,15 +4,27 @@
 from dagster import EventLogEntry
```
Comment (johannkm) [inline python_modules/dagster/dagster/_core/storage/asset_check_execution_record.py:6]: here's the new enums

**Comment by johannkm** [comment]
Added some checks that raise exceptions to toys, and double checked the UI:

Screenshot 2023-08-28 at 7 02 09 PM

We should show timestamp here (especially since the timestamp holds the link to the run), but will figure that out in a followup with @salazarm

File: python_modules/dagster/dagster/_core/storage/asset_check_execution_record.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/dagster/dagster/_core/storage/asset_check_execution_record.py:14]: Maybe `AssetCheckExecutionRecordStatus` would be a better name? It's possible that in some other context we'd want to store the resolved status.

File: python_modules/dagster/dagster/_core/storage/asset_check_execution_record.py [hunk 4]
```diff
@@ -4,15 +4,27 @@
...
 
 
-class AssetCheckExecutionStatus(enum.Enum):
+# We store a limit set of statuses in the database, and then resolve them to the actual status
+# at read time. This is because the write path is to store a planned event (which creates a row
+# with PLANNED status) then update the row when we get the check result. But if the check never
+# runs, the row stays in PLANNED status.
+class AssetCheckExecutionRecordStatus(enum.Enum):
     PLANNED = "PLANNED"
     SUCCESS = "SUCCESS"
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/storage/asset_check_execution_record.py:13]: Should this be SUCCEEDED to be consistent with FAILED? Also, same with EXECUTION_FAILURE below.

**Review by sryza — APPROVED** [review]
One more naming change comment to be consistent with the prior naming changes. After that, looks good to go.

**Pull request merged**